server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeDataLoader.java

```diff
@@ -0,0 +1,316 @@
+import static org.sonar.server.measure.ws.SnapshotDtoToWsPeriods.snapshotToWsPeriods;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.DEPRECATED_PARAM_BASE_COMPONENT_ID;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.PARAM_BRANCH;
+
+public class ComponentTreeDataLoader {
+  private static final Set<String> QUALIFIERS_ELIGIBLE_FOR_BEST_VALUE = ImmutableSet.of(Qualifiers.FILE, Qualifiers.UNIT_TEST_FILE);
+  private static final Joiner COMA_JOINER = Joiner.on(", ");
+
+  private final DbClient dbClient;
+  private final ComponentFinder componentFinder;
+  private final UserSession userSession;
+  private final ResourceTypes resourceTypes;
+
+  public ComponentTreeDataLoader(DbClient dbClient, ComponentFinder componentFinder, UserSession userSession, ResourceTypes resourceTypes) {
+    this.dbClient = dbClient;
+    this.componentFinder = componentFinder;
+    this.userSession = userSession;
+    this.resourceTypes = resourceTypes;
+  }
+
+  ComponentTreeData load(ComponentTreeRequest wsRequest) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      ComponentDto baseComponent = loadComponent(dbSession, wsRequest);
+      checkPermissions(baseComponent);
+      Optional<SnapshotDto> baseSnapshot = dbClient.snapshotDao().selectLastAnalysisByRootComponentUuid(dbSession, baseComponent.projectUuid());
+      if (!baseSnapshot.isPresent()) {
+        return ComponentTreeData.builder()
+          .setBaseComponent(baseComponent)
+          .build();
+      }
+
+      ComponentTreeQuery componentTreeQuery = toComponentTreeQuery(wsRequest, baseComponent);
+      List<ComponentDto> components = searchComponents(dbSession, componentTreeQuery);
+      List<MetricDto> metrics = searchMetrics(dbSession, wsRequest);
+      Table<String, MetricDto, Measure> measuresByComponentUuidAndMetric = searchMeasuresByComponentUuidAndMetric(dbSession, baseComponent, componentTreeQuery,
+        components,
+        metrics);
+
+      components = filterComponents(components, measuresByComponentUuidAndMetric, metrics, wsRequest);
+      components = sortComponents(components, wsRequest, metrics, measuresByComponentUuidAndMetric);
+
+      int componentCount = components.size();
+      components = paginateComponents(components, wsRequest);
+
+      return ComponentTreeData.builder()
+        .setBaseComponent(baseComponent)
+        .setComponentsFromDb(components)
+        .setComponentCount(componentCount)
+        .setMeasuresByComponentUuidAndMetric(measuresByComponentUuidAndMetric)
+        .setMetrics(metrics)
+        .setPeriods(snapshotToWsPeriods(baseSnapshot.get()))
+        .setReferenceComponentsByUuid(searchReferenceComponentsById(dbSession, components))
+        .build();
+    }
+  }
+
+  private ComponentDto loadComponent(DbSession dbSession, ComponentTreeRequest request) {
+    String componentKey = request.getComponent();
+    String componentId = request.getBaseComponentId();
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "getBaseComponentId"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)